src/Servers.jl

```diff
@@ -69,15 +69,20 @@ function check_rate_limit(tcp, rate_limit::Rational{Int})
 end
 
 "Convenience object for passing around server details"
-struct Server{S <: Union{SSLConfig, Nothing}, I <: Base.IOServer}
+struct Server{S <: Union{SSLConfig, Nothing}, I <: Base.IOServer, F <: Union{Function, Vector{Function}, Nothing}}
     ssl::S
     server::I
     hostname::String
     hostport::String
+    on_shutdown::F
 end
 
 Base.isopen(s::Server) = isopen(s.server)
-Base.close(s::Server) = close(s.server)
+Base.close(s::Server) = (shutdown(s.on_shutdown); close(s.server))
+
+shutdown(fns::Vector{Function}) = shutdown.(fns)
```

Comment:
```suggestion
shutdown(fns::Vector{Function}) = foreach(shutdown, fns)
```

Reply:
Nice, didn't know about this function, thanks!

Reply:
The reason is that `shutdown.(fns)` will allocate an array for the results, but you just want to call them and discard the result.

Reply:
Yeah that makes sense, I've used broadcasting and been slightly frustrated by the allocation before, so good to know about `foreach`